Genesis Commit For Heroku Multi Deploy Script

HoyaBoya opened the pull request:
This is the genesis commit for a deploy tool that can send a git slug to multiple Heroku Apps. I started baking it in BA.com, but then decided to make it a gem to get rid of some Rails crutches and have a `exe` outside of the app.

As an independent project:
1) we can drop it anything (WMS, etc etc)
2) OSS this later (there's nothing else doing this, I checked)

The actual code flow works like this. It's essentially an over-glorified Ruby wrapper around `git push heroku_remote git_reference`.

1) Get the environment and git reference we want to send
2) Take a look at the `dual_burner.yml` to see what the environment resolved to in terms of Heroku apps.
3) For every application, get the remote alias in a user's git config via Regexp. It can be named anything by a developer.
4) If everything looks good, prompt the user for a confirmation (learned some hard lessons from CSV importing)
5) If the challenge is fine, then run the `git push environment git_reference` in parallel. The output is captured and color coded so people can the difference.
6) Join all threads, grab a soda, and wait.

The test run with the gem installed locally with the following config in a BA.com repo yielded the following output.

```
fennel:
  - app_name: 'blueapron-fennel'
    name: 'Fennel'
    color: 'green'
  - app_name: 'blueapron-fennel'
    name: 'Fennel 2'
    color: 'blue'
staging:
  - app_name: 'blueapron-fennel'
    name: 'Fennel'
    color: 'blue'
production:
  - app_name: 'DO NOT DO'
    name: 'DO NOT DO'
    color: 'green'
```

<img width="946" alt="screen shot 2016-02-23 at 2 45 22 pm" src="https://cloud.githubusercontent.com/assets/134368/13264438/3f0a97ee-da3c-11e5-82de-eeec6316647c.png">

I'd also take a look at the README.md, to as well if anything is not explained well. If it's not explained well, then I should elucidate in the README.md.

Related Asana
https://app.asana.com/0/12264068303259/92713986846703

/cc @prsimp and @jackdwyer for DevOps review
/cc @midu @danielleadams @nickfarina @travisofthenorth @TheDudeWithTheThing @ipapas for Engineering review

midu commented on lib/dual_burner/git_config_parser.rb:
```diff
@@ -0,0 +1,32 @@
+    ##
+    # Turn an app name into a github heroku target.
+    def build_url(app:)
+      "git@heroku.com:#{app}.git"
```
heroku now uses `https://` for their remote URLs by default

midu commented on lib/dual_burner/deployer.rb:
```diff
@@ -0,0 +1,84 @@
+
+      challenge = "#{env}+#{heroku_apps.size}"
+      puts "If this is ok, please type in the following: #{challenge}"
+      response = STDIN.gets.chomp
```
how opening the diff in the user browser at least for prod deploys so they can visually go through the changes

HoyaBoya replied:
I'm a huge fan of `github.com/compare/`.

Let me see what we can do.

HoyaBoya commented:
@midu, I've addressed your HTTPS feedback here.

https://github.com/blueapron/dual_burner/commit/fc05096881147affee950ff76525c2da8541f902

midu commented on exe/rails_deploy:
```diff
@@ -0,0 +1,35 @@
+# The rest of arguments is the git reference
+# e.g rails_deploy staging master --force
+env = ARGV.shift
+git_reference = ARGV.join(' ')
```
may I interest you in [`OptionParser`](http://docs.ruby-lang.org/en/2.1.0/OptionParser.html)

HoyaBoya replied:
You may. Let's get it going.

midu commented on README.md:
````diff
@@ -1 +1,62 @@
+Add this line to your application's Gemfile:
+
+```ruby
+gem 'dual_burner'
````
I think you can do this here since it's only used as a binary: 

`gem 'dual_burner', require: false`

midu commented:
looks to me like a great first pass. We can start thinking about defaults (like default to using the current branch name when deploying to a preview env), but definitely not required for a first version. 

:+1: but would probably be a good idea to get another :+1: or two

HoyaBoya commented:
Thanks @midu. Let me get the OptionParser feedback incorporated.

@alexQueue @travisofthenorth do you have any thoughts on this as well? I'd love to incorporate your feedback before opening this up to BA.com folks, whom will be impacted by the cluster tool first.

HoyaBoya commented:
@midu and @prsimp finally got Opt Parse to work a bit better.

```
Kitchen-of-SunDawg:wms-server csun$ dual_burner_deploy -h
Usage: dual_burner_deploy [options]
    -i, --interactive [BOOLEAN]      Optional: Enable or disable user prompts.
    -g, --git-config [FILE]          Optional: The location of your git config.
    -b, --burner-config [FILE]       Optional: The location of your dual_burner.yml (if not config/dual_burner.yml).
    -r, --git-reference [STRING]     Required: The git reference to deploy.
    -e, --env [STRING]               Required: The git reference to deploy.
    -v, --version                    Show version
    -h, --help                       Displays help.
```

I've also cleaned some stuff up. Here's a sample interaction deploying to WMS Staging 4 and 5.

<img width="929" alt="screen shot 2016-02-24 at 12 16 55 pm" src="https://cloud.githubusercontent.com/assets/134368/13294194/d57e9d56-daf0-11e5-8f1a-d341f92d60c6.png">

HoyaBoya commented:
@jackdwyer @travisofthenorth @alexQueue @midu,

I've added a `dual_burner_last_release` helper to check commit hashes. It was a simple exercise to see how this might refactor out, and because, I'm punting on the compare for right now on #2.

<img width="831" alt="screen shot 2016-02-24 at 2 24 00 pm" src="https://cloud.githubusercontent.com/assets/134368/13298156/9fd32556-db03-11e5-91d5-e502ad204bc0.png">

<img width="824" alt="screen shot 2016-02-24 at 2 24 37 pm" src="https://cloud.githubusercontent.com/assets/134368/13298159/a2c18d3e-db03-11e5-8a26-fd746b4242b5.png">

midu commented on dual_burner.gemspec:
```diff
@@ -0,0 +1,37 @@
+
+  spec.files         = `git ls-files -z`.split("\x0").reject { |f| f.match(%r{^(test|spec|features)/}) }
+  spec.bindir        = 'exe'
+  spec.executables   = ['dual_burner_deploy', 'dual_burner_last_release']
```
shouldn't this be `exe/dual_burner_deploy, exe/dual_burner_deploy`?

or just `spec.executables   = spec.files.grep(%r{^exe/}) { |f| File.basename(f) }`

HoyaBoya replied:
`spec.executables = spec.files.grep(%r{^exe/}) { |f| File.basename(f) }` only returned the first file.

I can verify that what we have now works, since `gem build` explodes when it cannot find the referenced files.

midu commented on lib/dual_burner/base_command.rb:
```diff
@@ -0,0 +1,57 @@
+      end
+
+      result
+    end
```
what happens when the command returns something different than `0`? does it raise an exception? fails gracefully?

HoyaBoya replied:
For the deploy, it just gets printed out. Nothing is using the the process return value.

If in the context of deploy, something bad happens, you'll see it in the display messages. More interesting is if @jackdwyer would want to see something on his end.

I can return the code via `exit(N)` does that make sense?

HoyaBoya replied:
@midu I've taken a stab of interrogating error codes and returning `exit(0)` or `exit(1)` from the executable ruby scripts for @jackdwyer to use when he builds his tool to push without interaction.

HoyaBoya replied:
I will add, that we are unsure if that will work. We're testing the git server for our EC2 instances now...and for all intensive purposes, this is temporary work and may not live after we move off of Heroku.

jackdwyer replied:
Exit status of `1` should be sufficient if anything fails.  As we are running multiple deploys simultaneously, manual intervention will be needed if either/both fail.

midu commented on lib/dual_burner/last_release.rb:
```diff
@@ -0,0 +1,53 @@
+    end
+
+    def get_commit_hash(content:)
+      match = content.match(/Deploy ([A-Za-z0-9]{7})/)
```
Why not using git for that? Seems a little overkill to track every single output line just to be able to grep one of the last ones

```
git fetch remote_name
git rev-parse --short remote_name/master
```

HoyaBoya replied:
Oh that's neat and makes this not contingent on Heroku command line tools. Was too used to `heroku releases`...incorporating feedback.

HoyaBoya replied:
@midu I find that `ls-remote` does what I need a bit tidier. I'll be using that in the feedback commit.

```
Kitchen-of-SunDawg:wms-server csun$ git ls-remote git@heroku.com:ba-wms-staging-4.git HEAD
39099669e6a26d1de291091f25ab7f08709b2b87	HEAD
```

HoyaBoya commented:
@midu, I have removed `heroku releases` in favor of `ls-remote`. Thanks for the suggestion. Much much cleaner!

midu commented on exe/dual_burner_deploy:
```diff
@@ -0,0 +1,61 @@
+  optparse.parse!
+  required_args = [:env]
+  missing = required_args.select{ |param| options[param].nil? }
+  unless missing.empty?
```
`if missing.any?` would be a lot clearer to someone like me who has a hard time with double negatives

danielleadams replied:
I second this.

HoyaBoya replied:
Thanks, let's change to that.

midu commented on exe/dual_burner_deploy:
```diff
@@ -0,0 +1,61 @@
+rescue OptionParser::InvalidOption, OptionParser::MissingArgument
+  puts $!.to_s
+  puts optparse
+  exit
```
maybe introduce a `show_usage_and_exit` method?

HoyaBoya replied:
I think we can DRY up some of the work across the two `exe`. I will address.

midu commented on README.md:
````diff
@@ -1 +1,96 @@
+  - heroku_app_name: 'blueapron-production-www'
+    name: 'WWW'
+    color: 'purple'
+```
````
I don't know what others think, but at this point I don't think it would be unreasonable to enforce the same remote names for all the engineers. We could make the transition easy by offering a setup rake task in BA.

It would remove the need to parse the .git/config file, and it would make setting up BA 1% easier. The config could then look like this:

```yml
production:
  blueapron-api:
    remote: blueapron-api
    other_config_key_1: 'abc'
  blueapron:
    remote: blueapron
    other_config_key_1: 'def'
staging:
  blueapron-api:
    remote: blueapron-api-staging
    other_config_key_1: 'abc'
  blueapron:
    remote: blueapron-app-staging
    other_config_key_1: 'def'
fennel
  blueapron:
    remote: blueapron-fennel
# ...
```

Definitely not blocking, just a suggestion

jackdwyer replied:
I say enforce same remote names across the board.

HoyaBoya replied:
> I don't know what others think, but at this point I don't think it would be unreasonable to enforce the same remote names for all the engineers. 

Thanks @midu, as discussed offline, I am trying to lower the friction for on-boarding to this tool and the amount of mud I track into a developer's home. Forcing a convention would be a bit of dirt coming in through the door.

I'm happy to make this an enhancement request  but I'd prefer not to take action on a task to force generate a remote standard into people's `.git/config`.

#6

midu commented on lib/dual_burner/deployer.rb:
```diff
@@ -0,0 +1,69 @@
+      results = []
+
+      apps.each do |app|
+        remote = git_config_parser.find_remote(app: app['heroku_app_name']) || app['remote']
```
where does this `app['remote']` come from?

HoyaBoya commented:
@midu and @danielleadams, feedback addressed. I have moved a lot of the exe common logic into a module.

/cc @jackdwyer

HoyaBoya commented:
@alexQueue @travisofthenorth @dylandrop and @sclm and feedback here. I'd love to get some movement today so I can start working on the BA.com side.

/cc @jackdwyer

HoyaBoya commented on lib/dual_burner/base_command.rb:
```diff
@@ -0,0 +1,116 @@
+      @git_config_parser = DualBurner::GitConfigParser.new(configuration: configuration)
+      @git_config_parser.load
+
+      @apps = configuration_for
```
This emerged from a refactoring when all of the sub-classes of `BaseCommand` where invoking `configuration_for` too many times. Take a look at the sub-classes that use `apps`.

https://github.com/blueapron/dual_burner/pull/1/files#diff-e1382baff07028a896c42c9629911766R9

I will rename the method to be a bit more descriptive i.e. `get_apps_from_configuration` and be more defensive if the `env` is not present.

HoyaBoya replied:
You're not missing anything. This can be memoized too per your observation. This refactoring came pretty fast as I added a couple of more commands per our design review.

I will incorporate.

HoyaBoya commented on lib/dual_burner/base_command.rb:
```diff
@@ -0,0 +1,116 @@
+
+    ##
+    # Given an array of apps and values, determine if they are all consistent or not.
+    def verify_values(apps:, values:, context:)
```
It is. There are still more `apps:` references in the code base that can use the instance variable. I will change them per your feedback.

midu commented on spec/base_command_spec.rb:
```diff
@@ -0,0 +1,78 @@
+  - heroku_app_name: 'blueapron-staging-alpha'
+    name: 'Alpha'
+    color: 'blue'
+    remote: 'Endor'
```
The readme seems to be missing the `remote` key:

```yml
staging:
  - heroku_app_name: 'blueapron-staging-alpha'
    name: 'Alpha'
    color: 'blue'
  - heroku_app_name: 'blueapron-staging-bravo'
    name: 'Bravo'
    color: 'green'
```

HoyaBoya replied:
I will address. Was keeping this feature latent for @jackdwyer for his deploy to EC2 but I might as will be thorough.

jackdwyer replied:
Could we use `url` ?  Then we could forgo parsing the gitconfig

dual_burner would need to generate something like:
`git push git@github.com:blueapron/dual_burner master`

HoyaBoya replied:
I want to and it's part of @midu feature request at #6. It felt easier for people to see the commands executed on their machine similar to what they were doing.

Let me take a stab of removing the Git Parser in this PR. Having two people ask means we should do it.

prsimp commented on lib/dual_burner/deployer.rb:
```diff
@@ -0,0 +1,70 @@
+      # Show the user what we are about to do on their machine.
+      puts "These commands will be executed:"
+      apps.each do |app|
+        puts build_deploy_command(app: app, git_reference: git_reference).send(app.color).bold
```
Broken record here, but these kwargs seem to be totally superfluous.

prsimp commented on lib/dual_burner/git_config_parser.rb:
```diff
@@ -0,0 +1,36 @@
+
+    ##
+    # Find a developer's remote alias for the given heroku deploy target.
+    def find_remote(app:)
```
Not sure any of these `app:` kwargs are really necessary.

prsimp commented on README.md:
```diff
@@ -1 +1,112 @@
+
+You can verify that all versions are consistent with the following command.
+
+`dual_burner_last_release --env wms_testing`
```
I realize this is a first pass, but is there a reason we went with multiple individual commands vs. a command/sub-command approach (a la `git`)?

prsimp replied:
Also: is the eventual goal to support listing multiple releases, a la the `herokue releases` command?

HoyaBoya replied:
> I realize this is a first pass, but is there a reason we went with multiple individual commands vs. a command/sub-command approach (a la git)?

No real reason. This started as one script, then turned into a few based on feedback. It maybe born of trying to manage `ARGV` and `OptionParser` and making a CLI interface consistent.

If I had a bead on how to do it cleanly as one, I'll do it. Can I issue this for addressing later when I'm done with the current round of feedback?

prsimp commented on dual_burner.gemspec:
```diff
@@ -0,0 +1,30 @@
+
+  # Note: Files must be registered with git for things to package properly....
+  spec.files         = `git ls-files -z`.split("\x0").reject { |f| f.match(%r{^(test|spec|features)/}) }
+  spec.bindir        = 'exe'
```
I'm not totally sure, but I believe this is the Bundler scaffolding default as a nod to windows developers. I've never actually seen an `exe` directory in a Ruby gem.

IMO anything in `exe` should move to `bin` and these config directives should be updated accordingly.

HoyaBoya replied:
You are correct, the Bundler scaffold provided the `exe` directory with two util scripts in `bin`. If we moved to `bin` I'd have to exclude the `console` and `setup` from the Gem build.

Let me address other feedback first then think about this. I'm ambivalent but I don't want to fight against the Bundler scaffold, which I've done already to my detriment. This doesn't feel too egregiously bad, since the end result is the same, to drop 3 scripts into a person's `PATH`.

prsimp replied:
> I don't want to fight against the Bundler scaffold

I don't think this is fighting against it. It's job is done. I just think `bin` is more in keeping with the types of environments this will be deployed to. It's also the approach we've used in `blueapron-events`, `blueapron-error_notifier`, and `wms-client` -- AFAIK it was manually changed in all three instances.

HoyaBoya replied:
I'll make the change. Somewhere in the multi-verse, there's a Bizarro BA that uses Windows OS for dev machines  and there's a bizarro CSun that loves beer and is being told to keep this.

> I don't think this is fighting against it. It's job is done. I just think bin is more in keeping with the types of environments this will be deployed to

midu replied:
Seems like they did it to avoid conflicting bin paths between gems and projects using the gems, or something like that.

http://bundler.io/blog/2015/03/20/moving-bins-to-exe.html

HoyaBoya replied:
Actually @prsimp I'm going to decline to make this change. I've looked at both WMS Client and BA Events, and they don't have true executables, just utility scripts.

Those utility scripts are the same for this gem in `/bin` which is fine. But our Gem work has no precedent for the distribution of executables for the gem consumer. If we did, I'd certainly follow the established the pattern.

I've also taken the time to look at the reasons for the change in Bundler.

> This is a new convention. The current practice of both specifying bin/ as the executables directory and where we put binstubs and other development-only executables such as bin/rails or bin/setup, meant that Bundler-generated gems with executables were quite likely to have these development executables included in the built gem, and then installed along with the gem.

http://bundler.io/blog/2015/03/20/moving-bins-to-exe.html

This partition makes sense to me, since again, I don't want to conflate development and distribution scripts.

My apologies for doing a 180 here.

prsimp replied:
No worries. I didn't realize it was a bundler change, but their argument makes (enough) sense.

I am, of course, fine with following Bundler's convention. No changes needed.

prsimp commented on lib/dual_burner/deployer.rb:
```diff
@@ -0,0 +1,70 @@
+module DualBurner
+  class Deployer < DualBurner::BaseCommand
+    def deploy_to(git_reference: 'master')
```
This method name seems confusing. IMO `deploy_to` makes it seem as though the method expects a host to target for the deploy, not the git reference that is being deployed. In lieu of anything better, even just `deploy` seems more accurate.

(I also think this kwarg isn't useful, but you could probably guess that :wink:)

HoyaBoya replied:
You are correct. There used to be an `env` key-argument that was refactored out. That made the `deploy_to` more reasonable but now it's just stupid. I will address.

prsimp commented on lib/dual_burner/exe_helpers.rb:
```diff
@@ -0,0 +1,51 @@
+          exit(1)
+        end
+      rescue OptionParser::InvalidOption, OptionParser::MissingArgument
+        puts $!.to_s
```
Can we instead use the `rescue Foo => error` approach and use `error` vs. `$!`. If you really prefer using the globals, can we `require 'english'` and use `$ERROR_INFO`. IMO the Ruby shorthand globals are unnecessarily cryptic and despite its position alluding to what `$!` is, I don't think it's worth using when other options are readily available.

HoyaBoya replied:
I pulled in @jackdwyer PR at #3 without appreciating this feedback. I will incorporate.

prsimp commented on README.md:
```diff
@@ -1 +1,112 @@
+
+And then execute:
+
+  $ bundle
```
Needs either 4 leading spaces or the triple-backticks.

prsimp commented on README.md:
```diff
@@ -1 +1,112 @@
+
+Or install it yourself as:
+
+  $ gem install dual_burner
```
Same here re: 4 leading spaces or triple backticks.

prsimp commented on README.md:
````diff
@@ -1 +1,112 @@
+
+```
+[remote "origin"]
+  url = git@github.com:blueapron/dual_burner.git
````
This seems confusing -- if this is supposed to be included in an app's gemfile, then wouldn't it's `origin` point to somewhere other than `dual_burner`'s origin?

prsimp commented on README.md:
```diff
@@ -1 +1,112 @@
+
+#### Deploy of Master to Staging
+
+`dual_burner_deploy --env staging --git-reference master
```
Missing a closing backtick here.

prsimp commented on README.md:
```diff
@@ -1 +1,112 @@
+
+You can use a string for any git reference you would normally do.
+
+`dual_burner_deploy --env staging --git-reference 'master --force'`
```
Just appending `--force` to the git reference seems like a hack vs. having an actual `--force` option for `dual_burner`

HoyaBoya replied:
It is a hack. It was easier for me to take the reference as one big string vs trying to tackle every permutation of what we can do via git e.g. `master` `some_branch:master` `master --force`.

My goal was to not accidentally cripple underlying git CLI functionality with my wrapper CLI around it by not surfacing a `--X` or `--Y`.

I'm not sure I'll act on this feedback given these concerns. I don't want to provide a BA.com tool that is more frustrating to use the what they currently have.

prsimp commented on README.md:
```diff
@@ -1 +1,112 @@
+
+You can pass in file locations of your git configuration and deploy configuration if they are not under `.git` and `/config` respectively.
+
+`dual_burner_deploy --git-config .git/config --burner-config config/dual_burner.yml --interactive false --env staging --git-reference master`
```
Probably makes sense to use non-standard config file locations for this example.

prsimp commented:
Left a smattering of comments. Have we thought much about recovery strategies for partial deploy failure? Is the plan simply to have operators clean up the mess via a re-deploy if one of the deploys fails?

Also, as polish, I'd really love to see this become a command/sub-command style tool. In particular, I think `burn` (`burn deploy`, `burn config:set`, `burn config:get` etc.) is a nice, short command :) Of course this isn't something that needs to happen for this PR. My comment about the command/sub-command approach is really more of a question about whether it was considered and rejected for some reason.

dylandrop commented:
>  In particular, I think burn (burn deploy, burn config:set, burn config:get etc.) is a nice, short command

I support this too, looks more Heroku-ish so it's probably better for onboarding too.

HoyaBoya commented:
> I think burn (burn deploy, burn config:set, burn config:get etc.) is a nice, short command :)

> I support this, looks more Heroku-ish so it's probably better for onboarding too.

Like the way The Sith have a Rule of Two, my rule is that if 2 people I respect say the same thing, it is worth a try. I will try to unify the `exe`.

HoyaBoya commented:
@midu and @jackdwyer, I've incorporated feedback on *not* needing to parse `.git/config`.

@prsimp and @travisofthenorth, I've incorporated all of the style inconsistencies you've noticed with my half-ass refactoring.

HoyaBoya commented:
@prsimp and @dylandrop your feedback on a cleaner CLI is in #7. It will be merged into this PR if accepted.

HoyaBoya commented on lib/dual_burner/base_command.rb:
```diff
@@ -0,0 +1,114 @@
+        puts ""
+
+        for i in 0..(values.size - 1)
+          print apps[i].name.send(apps[i].color)
```
It's a bit of an assumption right now that N apps == N values, and that they are in the right order.

Let me take a stab to make this less idiotic.

HoyaBoya replied:
Cleaned up here @travisofthenorth: https://github.com/blueapron/dual_burner/commit/af7f35389d0a3071696f2e9928efd53da1ce6d70

HoyaBoya commented:
@travisofthenorth your feedback on that idiotic method to verify all values from an app cluster are the same has been cleaned up here.

https://github.com/blueapron/dual_burner/commit/af7f35389d0a3071696f2e9928efd53da1ce6d70

HoyaBoya commented:
Thanks for all the feedback everyone. I'll keep this open until Monday and if there are no further immediate concerns, try to get this in to move to the next phase of work.

I'll carry @midu's 3 day old approbation for moving forward.

https://github.com/blueapron/dual_burner/pull/1#issuecomment-187930480

Like a nice French triple-cream cheese (Pierre Robert or Epoisse), these :+1 s: just get better and better with age.

HoyaBoya commented:
@prsimp and @dylandrop, I spent a bit more time for Thor and namespace groups i.e. `config:set` and it seems like there is a bug with executables and namespace-ing. Hopefully we are content with `config_set` for now.

https://github.com/erikhuda/thor/issues/249#issuecomment-110679553

I'll create an issue for follow up on this if that is ok with you fine folks.

HoyaBoya commented:
This has been rebased to one. It is going into master, but please feel free to leave feedback or issues.

Thanks everyone!
